storage/src/main/java/org/apache/kafka/storage/internals/log/LogCleanerManager.java

```diff
@@ -0,0 +1,799 @@
+                        }
+                    }),
+                    metricTag
+            );
+
+            gaugeMetricNameWithTag
+                    .computeIfAbsent(UNCLEANABLE_BYTES_METRIC_NAME, k -> new ArrayList<>())
+                    .add(metricTag);
+        }
+
+        // a gauge for tracking the cleanable ratio of the dirtiest log
+        dirtiestLogCleanableRatio = 0.0;
+        metricsGroup.newGauge(MAX_DIRTY_PERCENT_METRIC_NAME, () -> (int) (100 * dirtiestLogCleanableRatio));
+
+        // a gauge for tracking the time since the last log cleaner run, in milliseconds
+        timeOfLastRun = Time.SYSTEM.milliseconds();
+        metricsGroup.newGauge(TIME_SINCE_LAST_RUN_MS_METRIC_NAME, () -> Time.SYSTEM.milliseconds() - timeOfLastRun);
+    }
+
+    public Map<String, List<Map<String, String>>> gaugeMetricNameWithTag() {
+        return gaugeMetricNameWithTag;
+    }
+
+    /**
+     * @return the position processed for all logs.
+     */
+    public Map<TopicPartition, Long> allCleanerCheckpoints() {
+        return inLock(lock, () -> checkpoints.values().stream()
+                .flatMap(checkpoint -> {
+                    try {
+                        return checkpoint.read().entrySet().stream();
+                    } catch (KafkaStorageException e) {
+                        LOG.error("Failed to access checkpoint file {} in dir {}",
+                                checkpoint.file().getName(), checkpoint.file().getParentFile().getAbsolutePath(), e);
+                        return Stream.empty();
+                    }
+                })
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+    }
+
+    /**
+     * Public for unit test. Get the cleaning state of the partition.
+     */
+    public Optional<LogCleaningState> cleaningState(TopicPartition tp) {
+        return inLock(lock, () -> Optional.ofNullable(inProgress.get(tp)));
+    }
+
+    /**
+     * Public for unit test. Set the cleaning state of the partition.
+     */
+    public void setCleaningState(TopicPartition tp, LogCleaningState state) {
+        inLock(lock, () -> inProgress.put(tp, state));
+    }
+
+    /**
+     * Choose the log to clean next and add it to the in-progress set. We recompute this
+     * each time from the full set of logs to allow logs to be dynamically added to the pool of logs
+     * the log manager maintains.
+     */
+    public Optional<LogToClean> grabFilthiestCompactedLog(Time time, PreCleanStats preCleanStats) {
+        return inLock(lock, () -> {
+            long now = time.milliseconds();
+            timeOfLastRun = now;
+            Map<TopicPartition, Long> lastClean = allCleanerCheckpoints();
+
+            List<LogToClean> dirtyLogs = logs.entrySet().stream()
+                    .filter(entry -> entry.getValue().config().compact &&
+                            !(inProgress.containsKey(entry.getKey()) || isUncleanablePartition(entry.getValue(), entry.getKey()))
```

Comment:
This is equivalent to the following, right? Avoiding an extra level of nesting probably makes the code a bit easier to understand.

```
                     !inProgress.containsKey(entry.getKey()) &&
                     !isUncleanablePartition(entry.getValue(), entry.getKey()
```